### ISSUE TITLE
HMC sampler returns repeated samples

### ISSUE THREAD

**Issue by oschulz** [bat/BAT.jl#142]
Currently, the HMC sampler seems to repeats samples the proposal are rejected in the MH step:

```julia
samples = bat_sample(PosteriorDensity(v -> LogDVal(1), NamedTupleDist(x = Normal())), 10^5, AHMC()).result
all(isone, samples.weight) == true
length(samples.v) > length(unique(samples.v.x))
```

If we get repeated samples from AdvancedHMC, we want to increase the weight of the sample in steps of one instead of repeating it in the output. AHMI, for example, needs unique samples.

**Comment by Cornelius-G** [comment]
Ok, so I had a look into this and think I know why this is and how it may be solved.
The only issue is that I now also have another problem when using AHMC sampler.
When trying to run the `plotting_examples.jl` with `AHMC()`, I get the following error:
```
ERROR: Density evaluation failed at v = (a = -3.2724767718471353, b = [-4.894893433967544, -9.461767430855591]) due to exception TypeError(:typeassert, "", Float64, Dual{Nothing}(-19.11897323345669,-4.671891130661713,-0.0,-0.0))
Stacktrace:
 [1] eval_density_logval(::BAT.GenericDensity{var"#7#8"}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{BAT.var"#logval_posterior#184"{PosteriorDensity{BAT.GenericDensity{var"#7#8"},BAT.DistributionDensity{ValueShapes.NamedTupleDist{(:a, :b),Tuple{Normal{Float64},Product{Continuous,Uniform{Float64},Array{Uniform{Float64},1}}},Tuple{ValueShapes.ValueAccessor{ValueShapes.ScalarShape{Real}},ValueShapes.ValueAccessor{ValueShapes.ArrayShape{Real,1}}}},BAT.HyperRectBounds{Float64}},BAT.HyperRectBounds{Float64},ValueShapes.NamedTupleShape{(:a, :b),Tuple{ValueShapes.ValueAccessor{ValueShapes.ScalarShape{Real}},ValueShapes.ValueAccessor{ValueShapes.ArrayShape{Real,1}}}}}},Float64},Float64,3},1}, ::ValueShapes.NamedTupleShape{(:a, :b),Tuple{ValueShapes.ValueAccessor{ValueShapes.ScalarShape{Real}},ValueShapes.ValueAccessor{ValueShapes.ArrayShape{Real,1}}}}) at
```
This seems to happen when calling `density_logval` in the `AHMCIterator` ctor. @oschulz, I saw you made some changes a few days ago affecting the `density_logval`. Might this be related to these changes? I tried a few things but did not find any working solution for this.
I also noticed the following when using the posterior of the `plotting_examples.jl`:
```
BAT.density_logval(posterior, (a = -9.5, b = [-10.4, 7.7])) == -28.1
BAT.density_logval(posterior, [-9.5, 10.4, 7.7]) == -30.3
```
Is this expected behavior?

**Comment by oschulz** [comment]
> I saw you made some changes a few days ago affecting the density_logval

Actually, it should be safer now than before, not the other way round, hopefully ...
Could you check again if it works with a commit before those changes, e.g. 3d50332 ?

**Comment by oschulz** [comment]
Looks like the `likelihood` defined in "plotting_examples.jl" is not compatible with ForwardDiff in it's current form:

```julia
f = v -> likelihood((a = v[1], b = v[2:3])).value
X = rand(prior)
f(X)
ForwardDiff.gradient(f, X)
```

results in

```
ERROR: TypeError: in typeassert, expected Float64, got ForwardDiff.Dual{Nothing,Float64,3}
```

You can now easily check autodiff-compatibility (currently only for ForwardDiff) using

```julia
BAT.density_logvalgrad(posterior, rand(prior))
```

You should either get a gradient or an exception.

**Comment by oschulz** [comment]
The problem lies in `Distributions._mixlogpdf1` - that code is not compatible with dual numbers it doesn't propagate the type of the parameters properly. We should report a bug or make a bugfix PR.

**Comment by oschulz** [comment]
I just changed `density_logval` to use `rethrow` to preserve the stacktrace of the original exception, makes it easier to track down such problems.

**Comment by Cornelius-G** [comment]
Ah, ok I see. I just thought I had tested AHMC with the plotting examples, but maybe this was with a different likelihood or so.
Thanks for pointing to the error and adding the `rethrow`.

**Comment by Cornelius-G** [comment]
The original issue with the repeated samples should be fixed with PR #144.

**Comment by oschulz** [comment]
Thanks!
